src/rust/lib/flexer/src/group.rs

```diff
@@ -0,0 +1,256 @@
+        let dfa = nfa.to_dfa();
+
+        let expected_nfa = NFA {
+            states: vec![
+                State::epsilon_links(&[1]),
+                State::epsilon_links(&[2]),
+                State::links(&[(32..=32,3)]),
+                State::epsilon_links(&[4]),
+                State::epsilon_links(&[5,8]),
+                State::epsilon_links(&[6]),
+                State::links(&[(32..=32,7)]),
+                State::epsilon_links(&[8]),
+                State::epsilon_links(&[5,9]).named("group0_rule0"),
+                State::default(),
+            ],
+            alphabet: Alphabet::new(&[0, 32, 33]),
+        };
+        let expected_dfa = DFA {
+            alphabet: Alphabet::new(&[0, 32, 33]),
+            links: DFA::links(vec![
+                vec![I,1,I],
+                vec![I,2,I],
+                vec![I,2,I],
+            ]),
+            end_states: vec![
+                None,
+                Some(EndState{priority:3,name:"group0_rule0".into()}),
+                Some(EndState{priority:3,name:"group0_rule0".into()}),
+            ]
+        };
+        assert_eq!(nfa,expected_nfa);
+        assert_eq!(dfa,expected_dfa);
+    }
+
+    #[test]
+    fn test_letter_and_spaces() {
+        let     letter = Pattern::range('a'..='z');
+        let     spaces = Pattern::char(' ').many1();
+        let mut state  = Group::default();
+
+        state.add_rule(Rule{pattern:letter, callback:"".into()});
+        state.add_rule(Rule{pattern:spaces, callback:"".into()});
+
+        let nfa = state.to_nfa();
+        let dfa = nfa.to_dfa();
+
+        let expected_nfa = NFA {
+            states: vec![
+                State::epsilon_links(&[1,3]),
+                State::links(&[(97..=122,2)]),
+                State::epsilon_links(&[11]).named("group0_rule0"),
+                State::epsilon_links(&[4]),
+                State::links(&[(32..=32,5)]),
+                State::epsilon_links(&[6]),
+                State::epsilon_links(&[7,10]),
+                State::epsilon_links(&[8]),
+                State::links(&[(32..=32,9)]),
+                State::epsilon_links(&[10]),
+                State::epsilon_links(&[7,11]).named("group0_rule1"),
+                State::default(),
+            ],
+            alphabet: Alphabet::new(&[32,33,97,123]),
+        };
+        let expected_dfa = DFA {
+            alphabet: Alphabet::new(&[32,33,97,123]),
+            links: DFA::links(vec![
+                vec![I,1,I,2,I],
+                vec![I,3,I,I,I],
+                vec![I,I,I,I,I],
+                vec![I,3,I,I,I],
+            ]),
+            end_states: vec![
+                None,
+                Some(EndState{priority:4,name:"group0_rule1".into()}),
+                Some(EndState{priority:4,name:"group0_rule0".into()}),
+                Some(EndState{priority:4,name:"group0_rule1".into()}),
+            ]
+        };
+
+        assert_eq!(nfa,expected_nfa);
+        assert_eq!(dfa,expected_dfa);
+    }
+}
```

Comment:
could you please add also benchmarks here? It would be important to track performance form the ground up. And please see how benchmark change when you for example change the Vectors in Vectors into 2d array